### PR TITLE
5 yelp api integration search

### PR DESCRIPTION
**Pull request by Viktor-K**
**Test Plan**

Show subset data comes from Yelp API, handle Yelp API errors and show UI message for invalid searched name.

Screenshot 2020-11-12 at 16 08 49

Screenshot 2020-11-12 at 16 08 37

### CODE REVIEW

File: src/API/API.ts [hunk 1]
```diff
@@ -1,34 +1,71 @@
...
+  const description = failure(errors).join(' - ');
+  return { type: 'Decoding', description: description };
+};
+
+const extractResponseBody = (res: Response): Promise<{ status: number; body: any }> =>
+  res.json().then(body => ({ status: res.status, body }));
+
+const decodingSuccessResponse: (
+  res: unknown
+) => taskEither.TaskEither<AppErrors, YelpAPIResponseT> = res => {
```
Comment (veej) [inline src/API/API.ts:36]: minor: just as a guideline we're following on UCZ, we decided to import the ADT (in this case capital `TaskEither`) directly from the specific file (in this case `import { TaskEither } from "fp-ts/TaskEither";`) and use the lowercase `taskEither` (`import { taskEither } from "fp-ts";`) only for the "operations" on the ADT (like map/flatMap/fold/...).
In this way, you can avoid the repetitive `taskEither.TaskEither` for the type.

File: src/components/ErrorMsg/ErrorMsg.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+import View from '../View/View';
+import { FormattedMessage } from 'react-intl';
+import { AppErrors, YelpError } from '../../API';
+
+import './errormsg.scss';
+
+export function ErrorMsg(props: { content: AppErrors; location: string }) {
+  const getMsgIdFromYelpApiErrorCode = (error: YelpError): string =>
+    error.code ? `YelpApiErrorCode.${error.code}` : 'YelpApiErrorCode.GENERIC';
```
Comment (veej) [inline src/components/ErrorMsg/ErrorMsg.tsx:10]: according to your codec, `error.code` can't be null/undefined. Can you receive an empty string as a code?

File: src/components/ErrorMsg/ErrorMsg.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          }}
+        />
+      )
+    }
+  ];
+  const renderErrMsg = () =>
+    errTypeMapper.reduce(
+      (prev, defiendType) => (props.content.type == defiendType.type ? defiendType.content : prev),
+      <FormattedMessage id="SearchContainer.error.generic" />
+    );
```
Comment (veej) [inline src/components/ErrorMsg/ErrorMsg.tsx:31]: this is really really strange. What are you trying to accomplish here?
Your `content.type` is a finite set of error types. If you define `errTypeMapper` as a `Record`, the type system will guarantee you always have an element for every error type and you won't need to use that `reduce` on the defined error types

File: src/components/ErrorMsg/ErrorMsg.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  return <View className="message-container animate-msg">{renderErrMsg()}</View>;
```
Comment (veej) [inline src/components/ErrorMsg/ErrorMsg.tsx:33]:
```suggestion
  return <View className={cx("message-container", "animate-msg")}>{renderErrMsg()}</View>;
```

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx [hunk 5]
```diff
@@ -1,27 +1,71 @@
 import * as React from 'react';
-import View from '../View/View';
+import * as QR from 'avenger/lib/QueryResult';
```
Comment (veej) [inline src/components/SearchResturantContainer/SearchResturantContainer.tsx:2]: pedantic comment:
```suggestion
import * as queryResult from 'avenger/lib/QueryResult';
```

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx [hunk 6]
```diff
@@ -1,27 +1,71 @@
...
+import { WithQueries } from 'avenger/lib/react';
 import { SearchBar } from '../SearchBar/SearchBar';
-import { SearchResults } from '../SearchResults/SearchResults';
-import Divider from '../Diveder/Divider';
 import { FormattedMessage } from 'react-intl';
 
+// import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
```
Comment (veej) [inline src/components/SearchResturantContainer/SearchResturantContainer.tsx:7]: leftover?

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx [hunk 7]
```diff
@@ -1,27 +1,71 @@
...
 
 export function SearchResturantContainer() {
-  const performSearch = (location: string, range: number) =>
-    alert(`You want to search resturant in ${location} around ${range} km`);
+  const [params, setParams] = React.useState<{ location: string; range: number }>({
+    location: '',
+    range: 0
+  });
+
+  const performSearch = async (location: string, range: number) =>
```
Comment (veej) [inline src/components/SearchResturantContainer/SearchResturantContainer.tsx:23]: why do you need the `async` here?

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx [hunk 8]
```diff
@@ -1,27 +1,71 @@
...
+      <View grow>
+        <FormattedMessage id="SearchContainer.empty.list" />
+      </View>
+    ) : (
+      <WithQueries
+        queries={{ searchResturantQuery }}
+        params={{ searchResturantQuery: { location: params.location, range: params.range } }}
+        render={QR.fold(
+          provideLoader,
+          (error: AppErrors) => (
```
Comment (veej) [inline src/components/SearchResturantContainer/SearchResturantContainer.tsx:39]: you shouldn't need to specify the type here. Avoid unneeded types

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx [hunk 9]
```diff
@@ -1,27 +1,71 @@
...
+          res => (
+            <View column hAlignContent="center" grow>
+              <FormattedMessage
+                id="SearchContainer.search.term"
+                values={{
+                  location: `${params.location}`,
+                  total: res.searchResturantQuery.total
+                }}
+              />
+              <SearchResults />
```
Comment (veej) [inline src/components/SearchResturantContainer/SearchResturantContainer.tsx:51]: Is this `SearchResult` still mocked?

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx [hunk 10]
```diff
@@ -1,27 +1,71 @@
...
+    location: '',
+    range: 0
+  });
+
+  const performSearch = async (location: string, range: number) =>
+    setParams({ location: location, range: range });
+
+  const provideLoader = () => <View>Loading...</View>;
+
+  const provideResults = () =>
```
Comment (veej) [inline src/components/SearchResturantContainer/SearchResturantContainer.tsx:28]: can this `provideResults` be a separate component?
You could move all this logic inside `SearchResults`, and pass the search params as a prop

File: src/config.ts [hunk 11]
```diff
@@ -1,25 +1,27 @@
...
   fold(
-    (errors) => {
-      throw new Error(["Invalid config provided:", failure(errors)].join("\n"));
+    errors => {
+      throw new Error(['Invalid config provided:', failure(errors)].join('\n'));
     },
-    (env) => ({
+    env => ({
       title: env.REACT_APP_TITLE,
+      yelpApiToken:
```
Comment (veej) [inline src/config.ts:23]: I don't understand why you put the token in the config if the value is still hardcoded. You should receive this token as an env variable and assign it here

File: src/model/yelp/yelp.tsx [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import * as t from 'io-ts';
+
+export const ResturantLocation = t.partial({
```
Comment (veej) [inline src/model/yelp/yelp.tsx:3]: as we discussed earlier, you should be more precise about the result you expect from the API. If you expect some field to be mandatory, you shouldn't use `t.partial` on the whole object.
We're using codecs to ensure that the APIs are compliant with a contract

File: src/model/yelp/yelp.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  code: t.string,
+  description: t.string
+});
+
+export const YelpAPIErrorResponse = t.type({
+  error: ErrMsg
+});
+
+export type YelpAPIResponseT = t.TypeOf<typeof YelpAPIResponse>;
+export type YelpAPIErrorResponseT = t.TypeOf<typeof YelpAPIErrorResponse>;
```
Comment (veej) [inline src/model/yelp/yelp.tsx:34]: please use the same name for the `io-ts` codec and its type:
```suggestion
export type YelpAPIResponse = t.TypeOf<typeof YelpAPIResponse>;
export type YelpAPIErrorResponse = t.TypeOf<typeof YelpAPIErrorResponse>;
```

**Review by veej — CHANGES_REQUESTED** [review]
Code is (almost) good. Left a few comments inline as usual.
Quite a big PR though. For the future PRs, consider splitting your work into even smaller PRs

File: src/components/SearchResturantContainer/SearchResturantContainer.tsx (hunk 9 above)
Comment (Viktor-K) [reply src/components/SearchResturantContainer/SearchResturantContainer.tsx:51]: Sure not I have delevop a separate component in another branch

**Review by veej — CHANGES_REQUESTED** [review]
Just a little comment about a useless cast. After that, we'll be ready to merge! 🚀

File: src/components/ErrorMsg/ErrorMsg.tsx [hunk 14]
```diff
@@ -2,33 +2,25 @@ import * as React from 'react';
...
-  ];
   const renderErrMsg = () =>
-    errTypeMapper.reduce(
-      (prev, defiendType) => (props.content.type == defiendType.type ? defiendType.content : prev),
-      <FormattedMessage id="SearchContainer.error.generic" />
+    props.content.type === 'YelpError' ? (
+      <FormattedMessage
+        id={getMsgIdFromYelpApiErrorCode(props.content as YelpError)}
+        values={{
+          msg: `${(props.content as YelpError).description}`,
```
[the remark this answers is not in the document]
Comment (Viktor-K) [reply src/components/ErrorMsg/ErrorMsg.tsx:17]: sure unuseful cast,  already inferred by the previous line, I'm overloaded of tasks and I forget something in some parts 😞

**Review by veej — APPROVED** [review]

**Pull request merged**